src/Compilers/VisualBasic/Portable/Lowering/Instrumentation/DynamicAnalysisInjector.vb

```diff
@@ -0,0 +1,303 @@
+﻿' Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+Imports Microsoft.CodeAnalysis.CodeGen
+Imports Microsoft.CodeAnalysis.VisualBasic.Symbols
+Imports Microsoft.CodeAnalysis.VisualBasic.Syntax
+Imports Roslyn.Utilities
+Imports System.Collections.Immutable
+Imports System.Diagnostics
+
+Namespace Microsoft.CodeAnalysis.VisualBasic
+
+    ''' <summary>
+    ''' This type provides means for instrumenting compiled methods for dynamic analysis.
+    ''' It can be combined with other <see cref= "Instrumenter" /> s.
+    ''' </summary>
+    Friend NotInheritable Class DynamicAnalysisInjector
+        Inherits CompoundInstrumenter
+
+        Private ReadOnly _method As MethodSymbol
+        Private ReadOnly _methodBody As BoundStatement
+        Private ReadOnly _createPayload As MethodSymbol
+        Private ReadOnly _spansBuilder As ArrayBuilder(Of SourceSpan)
+        Private _dynamicAnalysisSpans As ImmutableArray(Of SourceSpan) = ImmutableArray(Of SourceSpan).Empty
+        Private ReadOnly _methodEntryInstrumentation As BoundStatement
+        Private ReadOnly _payloadType As ArrayTypeSymbol
+        Private ReadOnly _methodPayload As LocalSymbol
+        Private ReadOnly _diagnostics As DiagnosticBag
+        Private ReadOnly _debugDocumentProvider As DebugDocumentProvider
+        Private ReadOnly _methodHasExplicitBlock As Boolean
+        Private ReadOnly _factory As SyntheticBoundNodeFactory
+
+        Public Shared Function TryCreate(method As MethodSymbol, methodBody As BoundStatement, factory As SyntheticBoundNodeFactory, diagnostics As DiagnosticBag, debugDocumentProvider As DebugDocumentProvider, previous As Instrumenter) As DynamicAnalysisInjector
+            Dim createPayload As MethodSymbol = GetCreatePayload(factory.Compilation, methodBody.Syntax, diagnostics)
+
+            ' Do not instrument the instrumentation helpers if they are part of the current compilation (which occurs only during testing). GetCreatePayload will fail with an infinite recursion if it Is instrumented.
```

Comment:
What part of the `If` condition checks for instrumentation helpers? `IsImplicitlyDeclared`? If so, that seems to catch more than just the instrumentation helpers so the comment may not be necessary.